### PR TITLE
fix(middleware-sdk-s3): process errors with 200 status code

### PR DESCRIPTION
**Pull request by siddsriv**
### Issue
Fixes: #3856

### Description
Process errors with 200 status code in throw200ExceptionsMiddleware

### Testing
Unit testing

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by trivikr** [comment]
The failing tests are integration tests in S3. It looks like there's some issue with mocking in those tests.

```console
  1) Throw 200 response
       should throw if UploadPartCopy() return with 200 and empty payload:

      AssertionError: expected promise to be rejected with an error including 'S3 aborted request' but got 'UnknownError'
      + expected - actual

      -UnknownError
      +S3 aborted request
      
  

  2) Throw 200 response
       should throw if CompleteMultipartUpload() return with 200 and empty payload:

      AssertionError: expected promise to be rejected with an error including 'S3 aborted request' but got 'UnknownError'
      + expected - actual

      -UnknownError
      +S3 aborted request
      
```

File: packages/middleware-sdk-s3/src/throw-200-exceptions.spec.ts [hunk 1]
```diff
@@ -15,6 +15,54 @@ describe("throw200ExceptionsMiddlewareOptions", () => {
     jest.clearAllMocks();
   });
 
+  describe("exceptions for code < 200 and >= 300", () => {
```
Comment (trivikr) [inline packages/middleware-sdk-s3/src/throw-200-exceptions.spec.ts:18]: nit: be explicit
```suggestion
  describe("exceptions for statusCode < 200 and >= 300", () => {
```

File: packages/middleware-sdk-s3/src/throw-200-exceptions.spec.ts [hunk 2]
```diff
@@ -15,6 +15,54 @@ describe("throw200ExceptionsMiddlewareOptions", () => {
...
+    mockStreamCollector.mockResolvedValue(Buffer.from(""));
+    mockUtf8Encoder.mockReturnValue("");
+    it("throws an exception if code is less than 200", async () => {
```
Comment (trivikr) [inline packages/middleware-sdk-s3/src/throw-200-exceptions.spec.ts:21]: Lot of code is shared between the two tests, where only statusCode is different.
You can use [it.each](https://jestjs.io/docs/api#testeachtablename-fn-timeout) for it as follows:

```js
it.each([199, 300])("throws for status code %d", async (statusCode) => {
```

File: packages/middleware-sdk-s3/src/throw-200-exceptions.spec.ts [hunk 3]
```diff
@@ -15,6 +15,54 @@ describe("throw200ExceptionsMiddlewareOptions", () => {
...
+          request: new HttpRequest({
+            hostname: "s3.us-east-1.amazonaws.com",
+          }),
+        });
+      } catch (e) {
+        expect(e).toBeDefined();
+        expect(e.name).toEqual("InternalError");
+        expect(e.message).toEqual("S3 aborted request");
+      }
+    });
```
Comment (trivikr) [inline packages/middleware-sdk-s3/src/throw-200-exceptions.spec.ts:42]: The middleware shouldn't throw an error when status code is <200 or >=300.

This code should be:
```suggestion
    const result = await handler({
        input: {},
        request: new HttpRequest({
          hostname: "s3.us-east-1.amazonaws.com",
        }),
      });
    expect(result.response).toBe(mockResponse);
```

Where:

```js
      mockNextHandler.mockReturnValue({
        response: mockResponse,
      });
```

**Review by trivikr — CHANGES_REQUESTED** [review]

**Comment by siddsriv** [comment]
(force) push above -- that refactors the tests. Needs more work before PR.

File: yarn.lock [hunk 4]
```diff
@@ -2080,9 +2080,9 @@
     universal-user-agent "^6.0.0"
 
 "@octokit/endpoint@^6.0.1":
```
Comment (trivikr) [inline yarn.lock:2082]: Undo changes in yarn.lock, as this file doesn't need to be edited.

File: packages/middleware-sdk-s3/src/throw-200-exceptions.spec.ts [hunk 5]
```diff
@@ -15,58 +16,81 @@ describe("throw200ExceptionsMiddlewareOptions", () => {
...
-    const handler = throw200ExceptionsMiddleware(mockConfig)(mockNextHandler, {} as any);
-    try {
-      await handler({
+
+    it.each([199, 300])("results for statusCode %i", async (statusCode) => {
+      mockNextHandler.mockReturnValue({
+        response: mockResponse,
+        statusCode,
+        // headers: {},
+        // body: "",
```
Comment (trivikr) [inline packages/middleware-sdk-s3/src/throw-200-exceptions.spec.ts:28]: Uncomment, and pass body with some content?
```suggestion
        headers: {},
        body: "body",
```

**Review by trivikr — CHANGES_REQUESTED** [review]

**Review by AllanZhengYP — APPROVED** [review]

**Review by kuhe — APPROVED** [review]

**Review by trivikr — APPROVED** [review]

**Pull request merged**